Core/src/org/sleuthkit/autopsy/datasourcesummary/datamodel/DataSourceInfoUtilities.java

```diff
@@ -133,9 +141,204 @@ static String getMetaFlagsContainsStatement(TSK_FS_META_FLAG_ENUM flag) {
+        if (maxCount < 0) {
+            throw new IllegalArgumentException("Invalid maxCount passed to getArtifacts, value must be at greater 0");
+        }
+
+        return createListFromMap(getArtifactMap(skCase, artifactType, dataSource, attributeType, sortOrder), maxCount);
+    }
+
     /**
      * Empty private constructor
      */
     private DataSourceInfoUtilities() {
     }
+    
+     /**
+     * Create a Map of lists of artifacts sorted by the given attribute.
+     *
+     * @param skCase        SleuthkitCase instance.
+     * @param artifactType  Type of artifacts to sort.
+     * @param dataSource    Data Source that the artifact belongs to.
+     * @param attributeType Attribute type to sort by.
+     * @param sortOrder     Sort order of the attributes, either ascending or
+     *                      descending.
+     *
+     * @return A Map of lists of artifacts sorted by the value of attribute
+     *         given type. Artifacts that do not have an attribute of the given
+     *         type will not be included.
+     *
+     * @throws TskCoreException
+     */
+    static private SortedMap<BlackboardAttribute, List<BlackboardArtifact>> getArtifactMap(SleuthkitCase skCase, BlackboardArtifact.Type artifactType, DataSource dataSource, BlackboardAttribute.Type attributeType, SortOrder sortOrder) throws TskCoreException {
+        SortedMap<BlackboardAttribute, List<BlackboardArtifact>> sortedMap = new TreeMap<>(new AttributeComparator(sortOrder));
+        List<BlackboardArtifact> artifactList = skCase.getBlackboard().getArtifacts(artifactType.getTypeID(), dataSource.getId());
+
+        for (BlackboardArtifact artifact : artifactList) {
+            BlackboardAttribute attribute = artifact.getAttribute(attributeType);
+            if (attribute == null) {
+                continue;
+            }
+
+            List<BlackboardArtifact> mapArtifactList = sortedMap.get(attribute);
+            if (mapArtifactList == null) {
+                mapArtifactList = new ArrayList<>();
+                sortedMap.put(attribute, mapArtifactList);
+            }
+
+            mapArtifactList.add(artifact);
+        }
+
+        return sortedMap;
+    }
+
+    /**
+     * Creates the list of artifacts from the sorted map and the given count.
+     *
+     * @param sortedMap Sorted map of artifact lists.
+     * @param maxCount  Maximum number of artifacts to return.
+     *
+     * @return List of artifacts, list will be empty if none were found.
+     */
+    static private List<BlackboardArtifact> createListFromMap(SortedMap<BlackboardAttribute, List<BlackboardArtifact>> sortedMap, int maxCount) {
+        List<BlackboardArtifact> artifactList = new ArrayList<>();
+
+        for (List<BlackboardArtifact> mapArtifactList : sortedMap.values()) {
+
+            if (maxCount == artifactList.size()) {
```

Comment:
I believe the intent is that if maxCount is zero, return all records.  It looks like this line causes no records to be returned in that event.